### PR TITLE
Avoid null terminating characters in strings from Utf8FromUtf16()

### PR DESCRIPTION
**Pull request by tgucio**
This PR adds passing source wchar_t* string length to WideCharToMultiByte() in Utf8FromUtf16() in order to avoid including the terminating null character in the resulting string. This was found when the Utf8FromUtf16() function in the generated utils.cpp file was used for another purpose (a platform specific extension using method channel).

From https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-widechartomultibyte:

> [in] cchWideChar
> Size, in characters, of the string indicated by lpWideCharStr. Alternatively, this parameter can be set to -1 if the string is null-terminated. If cchWideChar is set to 0, the function fails.
> If this parameter is -1, the function processes the entire input string, including the terminating null character. Therefore, the resulting character string has a terminating null character, and the length returned by the function includes this character.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (loic-sharma) [inline packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:56]: Should this be `source_length`?

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (loic-sharma) [inline packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:48]: Instead of using `wcslen`, could we substract 1 from the `target_length`?

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (loic-sharma) [inline packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:61]: Is this case possible?

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (tgucio) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:56]: Do you mean the if (target_length == 0...) condition? I think we want to check the resulting length (target_length) which will be 0 e.g. if there are invalid characters in an non-empty utf16_string (note WC_ERR_INVALID_CHARS is set).

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (tgucio) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:48]: Yes, this is the other approach but for some reason I thought getting source_length first looked more elegant.
But leaving -1 passed to WideCharToMultiByte() and subtracting 1 will avoid the extra counting. Let me change this.

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (tgucio) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:61]: Well, if we know the lengths of both the source string and resulting string we should pass them. But I'll change this to -1 and subtract 1 from target_length anyway as per the other comment above.

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (cbracken) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:56]: `target_length` is write -- WideCharToMultibyte returns the number of characters that are written to the `lpMultiByteStr` out param or, in the case where we pass null for that param, the number that are required. We then resize `utf8_string` (the target buffer) to that size and do the write.

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (cbracken) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:60]: See my comment above, you shouldn't need to subtract one. The output size is the number of bytes required for the buffer.

> Is this case possible?

It seems unlikely given that we just allocated a buffer the right size, but without knowing the implementation, checking the return value here seems like the right thing to do and it's an O(1) check, so pretty negligible compared to the encoding conversion that `WideCharToMultibyte` is doing.

**Review by cbracken — CHANGES_REQUESTED** [review]
I'm a bit concerned by this change in that it changes the semantics of these functions in a potentially breaking way for callers. I believe that in every use we have of these functions in the templates, we're assuming copying of the full string. 

There are cases where we ask users to regenerate the app project using flutter create (e.g. migrations) and changing the behaviour of these could break existing uses they have in their projects.

Can you provide a bit more motivation for this change?

**Comment by tgucio** [comment]
@cbracken to me the behaviour of adding the null terminating char to the resulting C++ `std::string` isn't intended.
Here's how I stumbled upon this: I recycled the `Utf8FromUtf16()` function in order to pass `wchar_t*` file path from `DragQueryFile()` to `channel_->InvokeMethod()` via `flutter::EncodableMap()`. To my amazement I found the paths in Dart had an extra character that couldn't be removed with trim() because it was a null character (which happens to be a valid UTF-8 code point).

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (tgucio) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:56]: Hmm, isn't std::string devoid of null terminating char? If we inlude the C string null char in the resulting std::string, it will become an zero UTF-8 code point.

**Comment by christopherfujino** [comment]
@cbracken or @loic-sharma do either of you have time to take another look at this one?

**Comment by christopherfujino** [comment]
> @cbracken or @loic-sharma do either of you have time to take another look at this one?

pinged on discord

**Comment by Jasguerrero** [comment]
Discussion happening on discord chat

File: packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp
Comment (cbracken) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp:56]: Ah I see - the issue is that we're actively including the nul-terminator as part of the string contents due to the length passed in and thus generating a C++ output string whose size is therefore one greater than it should be because of the extra nul-terminator byte. This makes sense. Apologies for the delay in getting back to this.

**Review by cbracken — APPROVED** [review]
lgtm for the fix. This is testable but we don't currently have a mechanism for testing template bugs. I'm going to play around with a means of doing so.

**Review by cbracken — COMMENTED** [review]
Just to be clear - please do not land this until we have a test in place. (I'm taking a look at doing so right now)

**Review by cbracken — DISMISSED** [review]
Alright after hacking in a manual test, this fix also failed but in a different way, instead we get `ERROR_INSUFFICIENT_BUFFER`. I do have a patched version of the fix that works -- passing in the input buffer size rather than letting `WideCharToMultiByte` compute it results in a pass and avoids the manual `target_length` adjustment.

I'll put together a patch that adds a mechanism to test this.

**Comment by gspencergoog** [comment]
Converting back to a draft PR until there are tests. (Desktop triage)

**Comment by cbracken** [comment]
@yaakovschectman has added infrastructure for testing template code. The tests for this code are in https://github.com/flutter/flutter/pull/118647. (Desktop triage)

**Comment by tgucio** [comment]
@cbracken Sure, I'll take a look at testing this. I'm guessing we'd only need a test in dev/integration_tests/windows_startup_test/test_driver/main_test.dart as #118646 takes care of ensuring the utils.cpp matches the current app template.

**Comment by loic-sharma** [comment]
@tgucio Yup that's right. I would update the test here: https://github.com/flutter/flutter/pull/118647/files#diff-9e45c9708d35a2474dcea00dc2aa3d74afcc6e7ecc832e062b9b470a990280d0R49-R58

**Comment by yaakovschectman** [comment]
Because you modify `packages/flutter_tools/templates/app_shared/windows.tmpl/runner/utils.cpp`, you will need to propagate the change to all tests that use that template file so that the file is up-to-date. In this case this means you will need to modify `sdk/dev/integration_tests/windows_startup_test/windows/runner/utils.cpp` and `sdk/dev/integration_tests/ui/windows/runner/utils.cpp`. I think you will also need to update the test linked in the above comment to account for the now-missing null codepoint.

**Comment by yaakovschectman** [comment]
I do not believe you tested the new function implementation before submitting the PR. You cannot pass -1 as `cchWideChar` to `WideCharToMultiByte` when `cbMultiByte` is not equal to the length of the string including the terminating null, or else it will fail with `ERROR_INSUFFICIENT_BUFFER`. `target_length` should be passed instead.

**Review by yaakovschectman — APPROVED** [review]
I've added tests and updated the template files. Can anyone else LGTM this before I merge it?

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by tgucio** [comment]
@cbracken 's change request was on outdated code - I've dismissed the review so it should unclog things. Thanks @yaakovschectman for the corrections and tests!

File: dev/integration_tests/windows_startup_test/lib/main.dart
Comment (cbracken) [inline dev/integration_tests/windows_startup_test/lib/main.dart:54]: The TODO doesn't look like it matches up with the string here.

**Review by cbracken — COMMENTED** [review]
lgtm with just one nitpick re: a comment.

**Pull request merged**
